### PR TITLE
CMR-7035 Added Associated DOIs for ECHO 10, DIF 10, UMM-C, indexed Associated DOIs with DOI, and added all translations.

### PR DESCRIPTION
**Pull request by eereiter**

### CODE REVIEW

File: umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj [hunk 1]
```diff
@@ -52,12 +52,42 @@
...
+                 :let [assoc-type (value-of assoc "gmd:associationType/gmd:DS_AssociationTypeCode")
+                       assoc-type (when assoc-type
+                                    (string/trim assoc-type))
+                       doi (iso-util/char-string-value assoc "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier/gmd:code")]
+                 :when (and doi (= "associatedDOI" assoc-type))]
+             {:DOI doi
+              :Title (iso-util/char-string-value assoc "gmd:aggregateDataSetName/gmd:CI_Citation/gmd:title")
+              :Authority (iso-util/char-string-value assoc (str "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier"
+                                                                "/gmd:authority/gmd:CI_Citation/gmd:citedResponsibleParty"
+                                                                "/gmd:CI_ResponsibleParty/gmd:organisationName"))}))]
```
Comment (mschmele) [inline umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj:91]: This feels like a lot for a `let`, but the function itself does so little that I'm not sure if it's worth putting this into its own function. If it's possible to clean it up that would be awesome, but I don't have any suggestions that I like, so I won't put up a fuss 🙂

File: indexer-app/src/cmr/indexer/data/concepts/collection.clj [hunk 2]
```diff
@@ -172,6 +172,17 @@
...
+(comment
+  (let [collection {:DOI {:DOI "A"}
+                    :AssociatedDOIs [{:DOI "b"}
+                                     {:DOI "C"}]}]
+   ;(mapv #(util/safe-lowercase %)
+    (into [(util/safe-lowercase (get-in collection [:DOI :DOI]))]
+    ;(concat [(get-in collection [:DOI :DOI])]
+      (mapv #(util/safe-lowercase (:DOI %)) (get collection :AssociatedDOIs)))))
+      ;(get-in collection [:DOI :DOI]))))
+      ;(mapcat :DOI (get collection :AssociatedDOIs)))))
```
[the remark this answers is not in the document]
Comment (eereiter) [reply indexer-app/src/cmr/indexer/data/concepts/collection.clj]: Oops - no. I have removed the comment block

File: umm-spec-lib/src/cmr/umm_spec/migration/version/core.clj [hunk 3]
```diff
@@ -61,7 +61,7 @@
   "Returns a sequence of version steps between begin and end, inclusive."
   [concept-type begin end]
   (->> (condp #(%1 %2) (customized-compare begin end)
-         neg?  (sort-by count (versions concept-type))
+         neg?  (versions concept-type)
```
[the remark this answers is not in the document]
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/migration/version/core.clj:64]: I have removed the non used current-version reference and I have changed the code as you suggested.

File: umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj (hunk 1 above)
Comment (eereiter) [reply umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj:91]: Hehe,  I don't think it is worth it either, so I will leave it.

**Review by mschmele — APPROVED** [review]

File: umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj (hunk 1 above)
Comment (jaybarra) [reply umm-spec-lib/src/cmr/umm_spec/xml_to_umm_mappings/iso_shared/doi.clj:91]: I think the simplest refactor to improve readability would be something like this
```
(defn assemble-doi
  [doi]
  {:DOI doi
   :Title (iso-util/char-string-value assoc "gmd:aggregateDataSetName/gmd:CI_Citation/gmd:title")
   :Authority (iso-util/char-string-value assoc (str "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier"
                                                     "/gmd:authority/gmd:CI_Citation/gmd:citedResponsibleParty"
                                                     "/gmd:CI_ResponsibleParty/gmd:organisationName"))})
(defn parse-associated-dois
  [doc associated-doi-xpath]
  (let [assocs
        (seq
         (for [assoc (select doc associated-doi-xpath)
               :let [assoc-type (value-of assoc "gmd:associationType/gmd:DS_AssociationTypeCode")
                     assoc-type (when assoc-type
                                  (string/trim assoc-type))
                     doi (iso-util/char-string-value assoc "gmd:aggregateDataSetIdentifier/gmd:MD_Identifier/gmd:code")]
               :when (and doi (= "associatedDOI" assoc-type))]
           (asseble-doi doi)))]))
           ```

**Review by jaybarra — APPROVED** [review]

**Pull request merged**